ledger-core/v2/virtual/handlers/vdelegatedrequestfinished.go

```diff
@@ -0,0 +1,179 @@
+	"github.com/insolar/assured-ledger/ledger-core/v2/insolar/payload"
+	"github.com/insolar/assured-ledger/ledger-core/v2/log"
+	"github.com/insolar/assured-ledger/ledger-core/v2/vanilla/injector"
+	"github.com/insolar/assured-ledger/ledger-core/v2/vanilla/throw"
+	"github.com/insolar/assured-ledger/ledger-core/v2/virtual/descriptor"
+	"github.com/insolar/assured-ledger/ledger-core/v2/virtual/object"
+)
+
+type SMVDelegatedRequestFinished struct {
+	// input arguments
+	Meta    *payload.Meta
+	Payload *payload.VDelegatedRequestFinished
+
+	objectSharedState object.SharedStateAccessor
+	objectReadyToWork smachine.SyncLink
+
+	// dependencies
+	objectCatalog object.Catalog
+}
+
+type stateIsNotReadyErrorMsg struct {
+	*log.Msg  `txt:"State is not ready"`
+	Reference string
+}
+
+type unExpectedVDelegateRequestFinished struct {
+	*log.Msg  `txt:"Unexpected VDelegateRequestFinished"`
+	Reference string
+	ordered   bool
+}
+
+var dSMVDelegatedRequestFinishedInstance smachine.StateMachineDeclaration = &dSMVDelegatedRequestFinished{}
+
+type dSMVDelegatedRequestFinished struct {
+	smachine.StateMachineDeclTemplate
+}
+
+func (*dSMVDelegatedRequestFinished) InjectDependencies(sm smachine.StateMachine, _ smachine.SlotLink, injector *injector.DependencyInjector) {
+	s := sm.(*SMVDelegatedRequestFinished)
+
+	injector.MustInject(&s.objectCatalog)
+}
+
+func (*dSMVDelegatedRequestFinished) GetInitStateFor(sm smachine.StateMachine) smachine.InitFunc {
+	s := sm.(*SMVDelegatedRequestFinished)
+	return s.Init
+}
+
+/* -------- Instance ------------- */
+
+func (s *SMVDelegatedRequestFinished) GetStateMachineDeclaration() smachine.StateMachineDeclaration {
+	return dSMVDelegatedRequestFinishedInstance
+}
+
+func (s *SMVDelegatedRequestFinished) Init(ctx smachine.InitializationContext) smachine.StateUpdate {
+	return ctx.Jump(s.stepGetObject)
+}
+
+func (s *SMVDelegatedRequestFinished) stepGetObject(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	s.objectSharedState = s.objectCatalog.GetOrCreate(ctx, s.Payload.Callee)
+
+	action := func(state *object.SharedState) {
+		s.objectReadyToWork = state.ReadyToWork
+	}
+
+	switch s.objectSharedState.Prepare(action).TryUse(ctx).GetDecision() {
+	case smachine.NotPassed:
+		return ctx.WaitShared(s.objectSharedState.SharedDataLink).ThenRepeat()
+	case smachine.Impossible:
+		ctx.Log().Fatal("failed to get object state: already dead")
+	case smachine.Passed:
+	default:
+		panic(throw.NotImplemented())
+	}
+
+	return ctx.Jump(s.awaitObjectReady)
+}
+
+func (s *SMVDelegatedRequestFinished) awaitObjectReady(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	if ctx.AcquireForThisStep(s.objectReadyToWork).IsNotPassed() {
+		return ctx.Sleep().ThenRepeat()
+	}
+
+	return ctx.Jump(s.stepProcess)
+}
+
+func (s *SMVDelegatedRequestFinished) stepProcess(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	setStateFunc := func(data interface{}) (wakeup bool) {
```

Comment:
I'd strongly recommend to use typed wrappers for shared data and avoid interface casting in SM code